scripts/tank_cmd.py

```diff
@@ -1230,7 +1254,148 @@ def run_engine_cmd(log, pipeline_config_root, context_items, command, using_cwd,
+    login/password values and the other for script/key values. Any other values
+    are ignored.
+
+    :param auth_path: Path to a file with credentials.
+
+    :returns: A tuple of list of key, value pairs for values parsed. For example,
+       ([("login", "user"), ("password", "12345")],
+        [("script", "name"), ("key", "12345")])
+       The first tuple is concerned with human user based authentication, the second
+       is concerned with script based authentication.
+
+    :raises InvalidCredentials: If the file doesn't exist, this exception is raised.
+    """
+    if not os.path.exists(auth_path):
+        raise InvalidCredentials("credentials file does not exist.")
+    # Read the dictionary from file
+    with open(auth_path) as auth_file:
+        file_data = yaml.load(auth_file)
+
+    # We've already fixed the problem of parsing a bunch of values on the
+    # command line that may be mixed, so reshuffle the information that way
+    # and read from credentials
+    fake_cmd_line = ["--%s=%s" % (arg, value) for arg, value in file_data.iteritems()]
 
+    return (_extract_args(fake_cmd_line, ARG_SCRIPT, ARG_KEY),
+            _extract_args(fake_cmd_line, ARG_LOGIN, ARG_PASSWORD))
+
+
+def _extract_credentials(cmd_line):
+    """
+    Finds credentials from the command-line if the user specified them and
+    removes them from the array.
+
+    :param cmd_line: Array of arguments passed to the tank command.
+
+    :returns: A tuple of (filtered command line arguments, credentials).
+    """
+
+    # Extract the credential pairs.
+    script_user_credentials = _extract_args(cmd_line, ARG_SCRIPT, ARG_KEY)
+    human_user_credentials = _extract_args(cmd_line, ARG_LOGIN, ARG_PASSWORD)
+    file_credentials_path = _extract_args(cmd_line, ARG_AUTH)
+
+    # If we have credentials in a file
+    if file_credentials_path:
+        # make sure nothing else was specified on the command line.
+        if script_user_credentials or human_user_credentials:
+            raise InvalidCredentials("can't mix command line credentials and file credentials.")
+        # Validate it's only been specified once.
+        _validate_only_once(file_credentials_path, ARG_AUTH)
+        # Read the credentials from file
+        script_user_credentials, human_user_credentials = _read_credentials_from_file(
+            file_credentials_path[0][1]
+        )
+
+    # If we have credentials for both user types.
+    if script_user_credentials and human_user_credentials:
+        raise InvalidCredentials("can't mix human user and script user credentials.")
+
+    if script_user_credentials:
+        _validate_args_cardinality(script_user_credentials, ARG_SCRIPT, ARG_KEY)
+        return dict(script_user_credentials)
+
+    if human_user_credentials:
+        _validate_args_cardinality(human_user_credentials, ARG_LOGIN, ARG_PASSWORD)
+        return dict(human_user_credentials)
+
+    # If no elements were specified, that's ok.
+    return None
```

Comment:
Looking at this code, it feels like we are extracting a bunch of stuff only to pass it to `ensure_authenticated`. I wonder if this method could just keep more of the details internally, and then maybe we don't need to pass lots of params to `ensure_authenticated`. To me this approach seems a little simpler. What do you think?


```python
def _extract_credentials(args):
    """
    Reads authentication options from the command line.
    If options exist and are correct, these are used to 
    authenticate the current session. 

    :returns: (authenticated, new_args), where authenticated is a boolean
              indicating that toolkit was authenticated based on the given 
              input arguments and new_args is the args input list with all 
              authenticated parameters removed.

    :raises: InvalidCredentials on error
    """

    authenticated = False

    (params, new_args) = _extract_args(args, [ARG_SCRIPT, ARG_KEY, ARG_LOGIN, ARG_PASSWORD, ARG_AUTH])

    if ARG_SCRIPT in params and ARG_LOGIN in params:
        raise InvalidCredentials("Please supply either script or login credentials.")

    if ARG_AUTH in params and (ARG_SCRIPT in params or ARG_LOGIN in params):
        raise InvalidCredentials("If you are using an external credentials file you cannot specify command line crednetials.")

    # login based authentication
    if ARG_LOGIN in params:
        login = params[ARG_LOGIN]
        password = params.get(ARG_PASSWORD)
        if password is None:
            raise InvalidCredentials("Please supply a --password parameter")
        # authenticate
        user = shotgun_auth.create_session_user(login=login, password=password)
        tank.set_current_user(user)
        authenticated = True

    if ARG_SCRIPT in params:
        script_user = params[ARG_SCRIPT]
        script_key = params.get(ARG_KEY)
        if script_key is None:
            raise InvalidCredentials("Please supply a --script-key parameter")
        # authenticate
        user = shotgun_auth.create_session_user(api_script=script_user, api_key=script_key)
        tank.set_current_user(user)
        authenticated = True

    if ARG_AUTH in params:
        (script_user, script_key) = _load_yaml(params[ARG_AUTH])
        # authenticate
        user = shotgun_auth.create_session_user(api_script=script_user, api_key=script_key)
        tank.set_current_user(user)
        authenticated = True
        
    return (authenticated, new_args)


```